### PR TITLE
Using Gated Residual Networks and Variable Selection Networks

### PR DESCRIPTION
**Pull request by ksalama**
This example demonstrates the use of Gated
Residual Network (GRN) and Variable Selection Networks (VSN), described [here](https://arxiv.org/abs/1912.09363), for structured data classification.

### CODE REVIEW

**Comment by ksalama** [comment]
Hello @fchollet - Kindly review is new example. Thanks :-)

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
+"""
+Title: Classification with Gated Residual and Variable Selection Networks
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/01/15
+Last modified: 2021/01/15
+Description: Using Gated Residual and Variable Selection Networks for income level prediction.
```
Comment (8bitmp3) [inline examples/structured_data/classification_with_grn_and_vsn.py:6]: Suggestion: comments similar to https://github.com/keras-team/keras-io/pull/341#discussion_r548923687 and https://github.com/keras-team/keras-io/pull/341#discussion_r548924150 regarding the title and description.

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+"""
+
+"""
+## Introduction
+
+This example demonstrates the use of Gated
+Residual Network (GRN) and Variable Selection Networks (VSN), proposed by
+[Bryan Lim et al.](https://arxiv.org/abs/1912.09363), for structured data
```
Comment (8bitmp3) [inline examples/structured_data/classification_with_grn_and_vsn.py:14]: Suggestions:

- Add the name of the paper ("Temporal Fusion Transformers for Interpretable Multi-horizon Time Series Forecasting by Bryan Lim et al..."), since it's not named after Gated Residual Networks (GRNs) and Variable Selection Networks (VSNs) your example is based on.
  - If your example is only about GRNs and not the TFT model (the main thesis in the paper you're referencing), this should be made clearer, I think. 
   
GRNs inside the TFT model:
image

- The TFT paper is from 2019 but Gated ResNets (GRNs) date back to at least 2018 (https://web.cse.ohio-state.edu/~wang.77/papers/TCW.icassp18.pdf) - can you check again that GRN/VSN were proposed in 2019 in the TFT paper?
- Variable selection with neural nets were apparently proposed back in 1996: https://www.sciencedirect.com/science/article/abs/pii/0925231295001212 - can you double check your source again?

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+classification. GRNs give the flexibility to the model to apply non-linear
+processing only where needed. VSNs allow the model to softly remove any
+unnecessary noisy inputs which could negatively impact performance.
+Together, those techniques help improving the learning capacity of deep neural
+network models.
+
+
+Note that this example should be run with TensorFlow 2.3 or higher.
```
Comment (8bitmp3) [inline examples/structured_data/classification_with_grn_and_vsn.py:22]: Suggestion:

- Try to avoid repetition ("Note that this example... [Headline]... This examples uses... ").

```suggestion
Note that to run the code you need to use TensorFlow 2.3 or higher.
```

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+"""
+## Prepare the data
+
+First we load the data from the UCI Machine Learning Repository to a pandas Dataframe.
```
Comment (8bitmp3) [inline examples/structured_data/classification_with_grn_and_vsn.py:51]: Suggestion:

- Lint everything before submitting the PR. It's "Pandas" not "pandas".

```suggestion
First we load the data from the UCI Machine Learning Repository to a Pandas Dataframe.
```

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+
+"""
+## The dataset
+
+This example uses the [United States Census Income
+Dataset](https://archive.ics.uci.edu/ml/datasets/Census-Income+%28KDD%29) provided
+by the [UC Irvine Machine Learning Repository](https://archive.ics.uci.edu/ml/index.php).
+The task is binary classification to determine whether a person makes over 50K a year.
+
+The dataset includes ~300K instances with 41 input features, 7 of which numerical,
```
Comment (8bitmp3) [inline examples/structured_data/classification_with_grn_and_vsn.py:33]: Suggestion:
 
```suggestion
The dataset includes ~300K instances with 41 input features, 7 of which are numerical,
```

**Review by 8bitmp3 — COMMENTED** [review]
A few suggestions (but again, I'm just a contributor):

- Add the name of the paper ("Temporal Fusion Transformers for Interpretable Multi-horizon Time Series Forecasting by Bryan Lim et al..."), since it's not named after Gated Residual Networks (GRNs) and Variable Selection Networks (VSNs) your example is based on.
  - If your example is only about GRNs and not the TFT model (the main thesis in the paper you're referencing), this should be made clearer, I think. 
 
GRNs inside the TFT model:
image

- The TFT paper is from 2019 but Gated ResNets (GRNs) date back to at least 2018 (https://web.cse.ohio-state.edu/~wang.77/papers/TCW.icassp18.pdf) - can you check again that GRN/VSN were proposed in 2019 in the TFT paper?
- Variable selection with neural nets were apparently proposed back in 1996: https://www.sciencedirect.com/science/article/abs/pii/0925231295001212 - can you double check your source again?
 
Cheers

File: examples/structured_data/classification_with_grn_and_vsn.py (hunk 2 above)
Comment (8bitmp3) [reply examples/structured_data/classification_with_grn_and_vsn.py:14]: I see that the TFT paper's authors claim to propose a GRN but they may not be the first ones, unless their version only shares the name.

The 2018 paper submitted for IEEE (https://web.cse.ohio-state.edu/~wang.77/papers/TCW.icassp18.pdf) even says: _"In this study, we have proposed a gated residual network with dilated convolutions, named GRN..."_

🤷

Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py]: @8bitmp3 - Thank you for your comments. I made it clear in the introduction that the example covers only GRNs an VSNs, rather than the whole TFT model described in the paper. I also added the paper title (the reason I wanted to avoid adding the paper name is that it can be confusing from the first sight as the paper is about time-series forecasting, unlike the example). I still prefer to put the author with a like to the paper, rather than the title.

Also, I checked the same references you shared with me as I was preparing the example :-) The TFT paper uses the same names but different architectures and applications of these models.

**Comment by ksalama** [comment]
@8bitmp3 - Thank you for your comments. I made it clear in the introduction that the example covers only GRNs an VSNs, rather than the whole TFT model described in the paper. I also added the paper title (the reason I wanted to avoid adding the paper name is that it can be confusing from the first sight as the paper is about time-series forecasting, unlike the example). I still prefer to put the author with a like to the paper, rather than the title.

Also, I checked the same references you shared with me as I was preparing the example :-) The TFT paper uses the same names but different architectures and applications of these models.

**Comment by ksalama** [comment]
@fchollet - a kind reminder to take a look at this PR. Thanks :-)

**Review by fchollet — COMMENTED** [review]
Thank you for the PR. Question: are Gated residual networks still relevant today? Aren't they superseded by Transformer?

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+[Gated Linear Units (GLUs)](https://arxiv.org/abs/1612.08083) provide the
+flexibility to suppress input that are not relevant for a given task.
+"""
+
+
+class GatedLinearUnit(layers.Layer):
+    def __init__(self, units):
+        super(GatedLinearUnit, self).__init__()
+        self.linear = layers.Dense(units)
+        self.sigmoid = layers.Dense(units, activation="sigmoid")
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py:284]: the `self.linear` and the `self.sigmoid` have two separate sets of weights. The `call` functions use both of them like this: `self.linear(inputs) * self.sigmoid(inputs)`. It is a kind of "self-gating" activation function.  See this [blog post](https://leimao.github.io/blog/Gated-Linear-Units/).

[the remark this answers is not in the document]
Comment (fchollet) [reply examples/structured_data/classification_with_grn_and_vsn.py:284]: Got it, I thought they were chained but they're actually multiplied. Thanks for the clarification

**Comment by ksalama** [comment]
@fchollet - You are right. The only difference, I think is that Transformers using (multi-headed) attentions, while GRN do not. I think GRN might be an interesting architecture for structured data.

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+The Gated Residual Network (GRN) works as follows:
+1. Applies the nonlinear ELU transformation to the inputs.
+2. Applies linear transformation followed by dropout.
+4. Applies GLU and adds the original inputs to the output of the GLU to perform skip
+(residual) connection.
+6. Applies layer normalization and produces the output.
+"""
+
+
+class GatedResidualNetwork(layers.Layer):
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py:302]: I tried to use the functional API to implement it, but I was stuck in defining the input shape:

```
def create_gated_residual_network(units, dropout_rate):
  inputs = keras.Input(shape=???)
  x1 = inputs
  x2 = layers.Dense(units, activation="elu")(x1)
  x2 = layers.Dense(units)(x2)
  x2 = layers.Dropout(dropout_rate)(x2)
  if x1.shape[-1] != self.units:
    x1 = layers.Dense(units)(x1)
  outputs = x1 + GatedLinearUnit(units)(x2)
  outputs = layers.LayerNormalization()(outputs)

  model = keras.Model(inputs=inputs, outputs=outputs)
  return model
```

I don't want to fix it to (None, N), and (None, None) doesn't work. Do you have any suggestions? I think implementing it as a `Layer` is not too bad :-)

**Comment by ksalama** [comment]
@fchollet - Any updates on this one? I think it is interesting to provide a slightly non-standard architecture for structured data, but let me know if you think it is not interesting enough :)

 Cheers

**Comment by ksalama** [comment]
@fchollet - Final call for this one. Shall I close the PR? :)

**Comment by fchollet** [comment]
> @fchollet - Final call for this one. Shall I close the PR? :)

I'll review it, sorry for the delay. It's been a busy period.

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+
+def encode_inputs(inputs, encoding_size):
+    encoded_features = []
+    for feature_name in inputs:
+        if feature_name in CATEGORICAL_FEATURES_WITH_VOCABULARY:
+            vocabulary = CATEGORICAL_FEATURES_WITH_VOCABULARY[feature_name]
+            # Create a lookup to convert a string values to an integer indices.
+            # Since we are not using a mask token nor expecting any out of vocabulary
+            # (oov) token, we set mask_token to None and  num_oov_indices to 0.
+            index = StringLookup(
```
[the remark this answers is not in the document]
Comment (fchollet) [reply examples/structured_data/classification_with_grn_and_vsn.py:253]: Like in the other example, I'm not sure how important this is. We could just benchmark the two and keep the layer in the model if it doesn't seem to create a performance bottleneck.

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/01/15
+Last modified: 2021/01/15
+Description: Using Gated Residual and Variable Selection Networks for income level prediction.
+"""
+
+"""
+## Introduction
+
+This example demonstrates the use of [Gated
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py:12]: Fixed

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+Residual Networks (GRN) and Variable Selection Networks (VSN), proposed by
+Bryan Lim et al. in [Temporal Fusion Transformers (TFT) for Interpretable
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py:14]: Done

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+Multi-horizon Time Series Forecasting]((https://arxiv.org/abs/1912.09363),
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py:15]: Fixed

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+"""
+## The dataset
+
+This example uses the [United States Census Income
+Dataset](https://archive.ics.uci.edu/ml/datasets/Census-Income+%28KDD%29) provided
+by the [UC Irvine Machine Learning Repository](https://archive.ics.uci.edu/ml/index.php).
+The task is binary classification to determine whether a person makes over 50K a year.
+
+The dataset includes ~300K instances with 41 input features, 7 of which are numerical,
+and the other 34 are categorical.
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py:39]: Done

File: examples/structured_data/classification_with_grn_and_vsn.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py:253]: Thank you for the tip. However, it doesn't see to be impacting the performance that much in this example. So I recommend keeping it as is.

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+## Create tf.data.Dataset for training and evaluation
+
+We create an input function to read and parse the file, and convert features and
+labels into a [`tf.data.Dataset`](https://www.tensorflow.org/guide/datasets) for
+training and evaluation.
+"""
+
+from tensorflow.keras.layers.experimental.preprocessing import StringLookup
+
+taget_label_lookup = StringLookup(
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py:179]: It seems that `StringLookup` is needed as we are looking up the target indices  inside the `tf.data` pipeline. Note that the the target 'strings' are received as a tensor of shape (batch_size,), and we need to convert the values in this tensor from strings (vocabulary) to indices, so I don't think a Python dict will be suffices...

Please correct me if I am wrong.

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+        if feature_name in CATEGORICAL_FEATURES_WITH_VOCABULARY:
+            features[feature_name] = tf.strings.join(
+                ["v=", tf.cast(features[feature_name], tf.dtypes.string)]
+            )
+    target_index = taget_label_lookup(target)
+    weight = features.pop(WEIGHT_COLUMN_NAME)
+    return features, target_index, weight
+
+
+def get_dataset_from_csv(csv_file_path, shuffle=False, batch_size=128):
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py:195]: I have now simplified the `process` function called in the `tf.data` pipeline, and added comments.

File: examples/structured_data/classification_with_grn_and_vsn.py [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+    optimizer=keras.optimizers.Adam(learning_rate=learning_rate),
+    loss=keras.losses.BinaryCrossentropy(),
+    metrics=[keras.metrics.BinaryAccuracy()],
+)
+
+print("Start training the model...")
+train_dataset = get_dataset_from_csv(
+    train_data_file, shuffle=True, batch_size=batch_size
+)
+model.fit(train_dataset, epochs=num_epochs)
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py]: I added a validation split and the `EarlyStopping` callback

The early stopping callback has `restore_best_weights=True` argument, which restores the best model. So, I don't think both `ModelCheckpoint` and `EarlyStopping`.

What do you think?

File: examples/structured_data/classification_with_grn_and_vsn.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/classification_with_grn_and_vsn.py:314]: The reason is that the GRN is used two times, with two different input shapes:
1. for each feature independently: the shape here will be the embedding size
2. for the concatenated feature embeddings: the shape here embedding size X num of features

I think we can workout the shape and pass it as a parameter.

So which implementation do recommend?

**Comment by ksalama** [comment]
@fchollet Please find my replies. I assume that the `Continuous integration / docker-image (pull_request)` check failure is nothing to do with my code, right?

**Comment by fchollet** [comment]
> It seems that StringLookup is needed as we are looking up the target indices inside the tf.data pipeline. Note that the the target 'strings' are received as a tensor of shape (batch_size,), and we need to convert the values in this tensor from strings (vocabulary) to indices, so I don't think a Python dict will be suffices...

I would recommend simply doing the target encoding in the pandas dataframe, rather than after it gets converted to tf.data. This simplifies the code a bit.

**Comment by ksalama** [comment]
@fchollet I moved the target preprocessing to the pandas dataframe, and generated the other files.

Cheers

**Review by fchollet — APPROVED** [review]
